File: app/src/main/java/com/b502/minedroid/utils/MapItem.java
package com.b502.minedroid.utils;

import android.graphics.drawable.Drawable;
import android.widget.Button;

import androidx.appcompat.widget.AppCompatButton;
import androidx.core.content.ContextCompat;


public class MapItem {
    public enum State {
        DEFAULT, OPENED, FLAGED, BOOM, MISFLAGED
    }

    //地图项
    boolean isMine;
    int mineCount;
    AppCompatButton viewButton;
    State buttonState = State.DEFAULT;

    public State getButtonState() {
        return buttonState;
    }

    public boolean isMine() {
        return isMine;
    }

    public int getMineCount() {
        return mineCount;
    }

    public Button getViewButton() {
        return viewButton;
    }

    public void setMine(boolean mine) {
        isMine = mine;
    }

    public void setMineCount(int mineCount) {
        this.mineCount = mineCount;
    }

    public void setViewButton(AppCompatButton viewButton) {
        this.viewButton = viewButton;
    }

    public MapItem(boolean ismine) {
        setMine(ismine);
    }

    public void setButtonState(MapItem.State state) {
        this.buttonState = state;
        if (this.viewButton != null) {
            if (state == MapItem.State.DEFAULT) {
                this.viewButton.setText("");
                Drawable drawable = ContextCompat.getDrawable(this.getViewButton().getContext(), androidx.appcompat.R.drawable.abc_btn_colored_material);
                this.viewButton.setBackground(drawable);
                int tmp = this.viewButton.getCompoundDrawablePadding()+15;
                this.viewButton.setPadding(tmp, tmp, tmp, tmp);
            } else if (state == MapItem.State.FLAGED) {
                this.viewButton.setText("标");
            } else if (state == MapItem.State.OPENED) {
                int mineCount = this.mineCount;
                if (mineCount != 0) {
                    this.viewButton.setText(Integer.toString(mineCount));
                }
                this.viewButton.setBackgroundColor(0x00000000);
            } else if (state == State.MISFLAGED) {
                this.viewButton.setText("X");
            } else if (state == State.BOOM) {
                this.viewButton.setText("雷");
            }
        }
    }
}

File: app/src/main/java/com/b502/minedroid/utils/MapItem.kt
package com.b502.minedroid.utils

import android.app.Activity
import androidx.appcompat.widget.AppCompatButton
import androidx.core.content.ContextCompat

class MapItem(context: Activity, mine: Boolean) {
    //地图项
    var mineCount: Int = if (mine) 10 else 9
        set(value) {
            field = value
            if (buttonState == State.OPENED) updateText()
        }
    var viewButton: AppCompatButton? = null
    var buttonState: State = State.DEFAULT
        set(value) {
            field = value
            updateText()
        }
    val x: Int? get() = (viewButton?.tag as IntArray?)?.get(0)
    val y: Int? get() = (viewButton?.tag as IntArray?)?.get(1)
    var isMine: Boolean
        get() = mineCount == 10
        set(v) {
            if (v) mineCount = 10
        }
    val drawable = ContextCompat.getDrawable(
        context, androidx.appcompat.R.drawable.abc_btn_colored_material
    )

    private fun updateText() {
        when (buttonState) {
            State.DEFAULT -> {
                viewButton?.text = ""
                viewButton?.background = drawable
                viewButton?.also { button ->
                    val tmp = button.compoundDrawablePadding + 15
                    button.setPadding(tmp, tmp, tmp, tmp)
                }
            }

            State.FLAGED -> {
                viewButton?.text = "标"
            }

            State.OPENED -> {
                if (mineCount != 0) {
                    viewButton?.text = mineCount.toString()
                }
                viewButton?.setBackgroundColor(0x00000000)
            }

            State.MISFLAGED -> {
                viewButton?.text = "X"
            }

            State.BOOM -> {
                viewButton?.text = "雷"
            }
        }
    }

    enum class State {
        DEFAULT, OPENED, FLAGED, BOOM, MISFLAGED
    }
}
